src/net/tcp/nbd.c

```diff
@@ -0,0 +1,1095 @@
+
+	/* Shut down interface */
+	intf_shutdown ( &nbdcmd->block, rc );
+
+	/* Drop session's reference */
+	ref_put ( &nbdcmd->refcnt );
+	nbd->command = NULL;
+}
+
+/** NBD command block interface operations */
+static struct interface_operation nbdcmd_block_op[] = {
+	INTF_OP ( intf_close, struct nbd_command *, nbdcmd_close ),
+};
+
+/** NBD command block interface descriptor */
+static struct interface_descriptor nbdcmd_block_desc =
+	INTF_DESC ( struct nbd_command, block, nbdcmd_block_op );
+
+/**
+ * Create NBD command
+ *
+ * @v nbd		NBD session
+ * @v block		Block data interface
+ * @v type		NBD command type
+ * @v lba		Starting logical block address
+ * @v count		Number of blocks to transfer
+ * @v buffer		Data buffer
+ * @v len		Length of data buffer
+ * @ret rc		Return status code
+ */
+static int nbd_command ( struct nbd_session *nbd, struct interface *block,
+			 enum nbd_command_type type,
+			 uint64_t lba, unsigned int count,
+			 userptr_t buffer, size_t len ) {
+	struct nbd_command *nbdcmd;
+
+	/* Sanity check */
+	assert ( nbd->command == NULL && nbd->rx_state == NBD_RX_TRANS_REP_CMD );
+
+	/* Allocate and initialise structure */
+	nbdcmd = zalloc ( sizeof ( *nbdcmd ) );
+	if ( ! nbdcmd )
+		return -ENOMEM;
+
+	ref_init ( &nbdcmd->refcnt, nbdcmd_free );
+	intf_init ( &nbdcmd->block, &nbdcmd_block_desc, &nbdcmd->refcnt );
+
+	nbdcmd->nbd = nbd;
+	nbd->command = nbdcmd;
+	ref_get ( &nbdcmd->nbd->refcnt );
+
+	nbdcmd->type = type;
+	nbdcmd->lba = lba;
+	nbdcmd->count = count;
+
+	nbdcmd->data_buffer = buffer;
+	nbdcmd->data_len = len;
+
+	/* Attach to parent interface, transfer reference to session,
+	 * and return
+	 */
+	intf_plug_plug ( &nbdcmd->block, block );
+	return 0;
+}
+
+
+/**
+ * Free NBD session
+ *
+ * @v refcnt		Reference counter
+ */
+static void nbd_free ( struct refcnt *refcnt ) {
+	struct nbd_session *nbd =
+		container_of ( refcnt, struct nbd_session, refcnt );
+
+	uri_put ( nbd->uri );
+	free ( nbd );
+}
+
+/**
+ * Shut down NBD interface
+ *
+ * @v nbd		NBD session
+ * @v rc		Reason for close
+ */
+static void nbd_close ( struct nbd_session *nbd, int rc ) {
+	/* A TCP graceful close is still an error from our point of view */
+	if ( rc == 0 )
+		rc = -ECONNRESET;
+
+	DBGC ( nbd, "NBD %p closed: %s\n", nbd, strerror ( rc ) );
+
+	/* Stop transmission process */
+	process_del ( &nbd->process );
+
+	if ( nbd->command )
+		nbdcmd_close ( nbd->command, rc );
+
+	/* Shut down interfaces */
+	intfs_shutdown ( rc, &nbd->block, &nbd->socket, NULL );
+}
+
+
+/****************************************************************************
+ *
+ * Block to NBD interface
+ *
+ */
+
+/**
+ * Check NBD flow-control window
+ *
+ * @v nbd		NBD session
+ * @ret len		Length of window
+ */
+static size_t nbd_block_window ( struct nbd_session *nbd ) {
+	DBGC ( nbd, "NBD %p block window\n", nbd );
+
+	if ( nbd->rx_state >= NBD_RX_TRANS_REP_CMD &&
+	     nbd->command == NULL )
+		/* We cannot handle concurrent commands */
```

Comment:
Please use braces for a multi-line `if` statement such as this

Reply:
OK, braces added